File: web_version/pmugg_files/viewport.js
ms.viewport = function(scale, centerX, centerY) {
	this.scale = scale;
	this.centerX = centerX;
	this.centerY = centerY;
	this.baseScale = scale;
	this.baseCenterX = centerX;
	this.baseCenterY = centerY;
	this.width  = ms.mainController.getCanvasWidth();
	this.height = ms.mainController.getCanvasHeight();
};

ms.viewport.prototype.setBaseScale = function(baseScale) {
	this.baseScale = baseScale;
};

ms.viewport.prototype.resize = function(w, h) {
	this.width = w;
	this.height = h;
};

ms.viewport.prototype.transformX = function(x) {
	return this.scale * (x - this.centerX) + this.width / 2;
};

ms.viewport.prototype.transformY = function(y) {
	return this.scale * (y - this.centerY) + this.height / 2;
};

ms.viewport.prototype.inverseTransform = function(screenPosition) {
	var x = (screenPosition.x - this.width  / 2) / this.scale + this.centerX;
	var y = (screenPosition.y - this.height / 2) / this.scale + this.centerY;
	return new ms.vec2(x, y);
};

ms.viewport.prototype.transform = function(position, offset, result) {
	var x = position.x;
	var y = position.y;
	if (offset) {
		x += offset.x;
		y += offset.y;
	}
	result.x = this.transformX(x);
	result.y = this.transformY(y);
};

ms.viewport.prototype.transform1 = function(position) {
	var result = new ms.vec2(0, 0);
	this.transform(position, ms.vec2.ORIGIN, result);
	return result;
};

ms.viewport.prototype.move = function(dx, dy) {
	this.centerX += dx / this.scale;
	this.centerY += dy / this.scale;
};

ms.viewport.prototype.zoom = function(dir, focusX, focusY) {
	var startScale = this.scale;
	var z;
	if (dir == 1) {
		z = ms.mainController.ZOOM_AMOUNT;
	} else {
		z = 1 / ms.mainController.ZOOM_AMOUNT;
	}
	this.scale *= z;
	if (this.scale < this.baseScale) {
		this.scale = this.baseScale;
		this.centerX = this.baseCenterX;
		this.centerY = this.baseCenterY;
	} else {
		var fx = this.centerX + (focusX - this.width / 2) / startScale;
		var fy = this.centerY + (focusY - this.height / 2) / startScale;
		this.centerX = fx + (this.centerX - fx) / z;
		this.centerY = fy + (this.centerY - fy) / z;
	}
};

File: cpp_version/pmugg/viewport.cpp
#include "viewport.h"
#include "main_controller.h"

namespace ms {

Viewport::Viewport(float scale, float centerX, float centerY)
    : scale(scale)
    , centerX(centerX)
    , centerY(centerY)
    , baseScale(scale)
    , baseCenterX(centerX)
    , baseCenterY(centerY)
    , width(MainController::getCanvasWidth())
    , height(MainController::getCanvasHeight()) {
}

void Viewport::setBaseScale(float baseScale) {
    this->baseScale = baseScale;
}

void Viewport::resize(int w, int h) {
    width = w;
    height = h;
}

float Viewport::transformX(float x) {
    return scale * (x - centerX) + width / 2;
}

float Viewport::transformY(float y) {
    return scale * (y - centerY) + height / 2;
}

Vec2 Viewport::inverseTransform(const Vec2& screenPosition) {
    float x = (screenPosition.x - width / 2) / scale + centerX;
    float y = (screenPosition.y - height / 2) / scale + centerY;
    return Vec2(x, y);
}

void Viewport::transform(const Vec2& position, const Vec2& offset, Vec2& result) {
    float x = position.x + offset.x;
    float y = position.y + offset.y;
    result.x = transformX(x);
    result.y = transformY(y);
}

Vec2 Viewport::transform1(const Vec2& position) {
    Vec2 result(0, 0);
    transform(position, Vec2(0, 0), result);
    return result;
}

void Viewport::move(float dx, float dy) {
    centerX += dx / scale;
    centerY += dy / scale;
}

void Viewport::zoom(int dir, float focusX, float focusY) {
    float startScale = scale;
    float z = (dir == 1) ? 
        MainController::ZOOM_AMOUNT : 
        1.0f / MainController::ZOOM_AMOUNT;
    
    scale *= z;
    
    if (scale < baseScale) {
        scale = baseScale;
        centerX = baseCenterX;
        centerY = baseCenterY;
    } else {
        float fx = centerX + (focusX - width / 2) / startScale;
        float fy = centerY + (focusY - height / 2) / startScale;
        centerX = fx + (centerX - fx) / z;
        centerY = fy + (centerY - fy) / z;
    }
}

} // namespace ms 